pepsico/projections/maproom.py

```diff
@@ -0,0 +1,255 @@
+import psycopg2
+from psycopg2 import sql
+import shapely
+from shapely import wkb
+from shapely.geometry.multipolygon import MultiPolygon
+from globals_ import FLASK, GLOBAL_CONFIG
+
+def register(FLASK, config):
+    PFX = f"{GLOBAL_CONFIG['url_path_prefix']}/{config['core_path']}"
+    TILE_PFX = f"{PFX}/tile"
+
+    # App
+
+    APP = dash.Dash(
+        __name__,
+        server=FLASK,
+        external_stylesheets=[
+            dbc.themes.BOOTSTRAP,
+        ],
+        url_base_pathname=f"{PFX}/",
+        meta_tags=[
+            {"name": "description", "content": "Forecast"},
+            {"name": "viewport", "content": "width=device-width, initial-scale=1.0"},
+        ],
+    )
+    APP.title = "Forecast"
+
+    APP.layout = layout.app_layout()
+
+    def adm_borders(shapes):
+        with psycopg2.connect(**GLOBAL_CONFIG["db"]) as conn:
+            s = sql.Composed(
+                [
+                    sql.SQL("with g as ("),
+                    sql.SQL(shapes),
+                    sql.SQL(
+                        """
+                        )
+                        select
+                            g.label, g.key, g.the_geom
+                        from g
+                        """
+                    ),
+                ]
+            )
+            df = pd.read_sql(s, conn)
+
+        df["the_geom"] = df["the_geom"].apply(lambda x: wkb.loads(x.tobytes()))
+        df["the_geom"] = df["the_geom"].apply(
+            lambda x: x if isinstance(x, MultiPolygon) else MultiPolygon([x])
+        )
+        shapes = df["the_geom"].apply(shapely.geometry.mapping)
+        for i in df.index: #this adds the district layer as a label in the dict
+            shapes[i]['label'] = df['label'][i]
+        return {"features": shapes}
+
+
+    def make_adm_overlay(
+        adm_name, adm_sql, adm_color, adm_lev, adm_weight, is_checked=False,
+    ):
+        border_id = {"type": "borders_adm", "index": adm_lev}
+        return dlf.Overlay(
+            dlf.GeoJSON(
+                id=border_id,
+                data=adm_borders(adm_sql),
+                options={
+                    "fill": True,
+                    "color": adm_color,
+                    "weight": adm_weight,
+                    "fillOpacity": 0
+                },
+            ),
+            name=adm_name,
+            checked=is_checked,
+        )
+
+
+    @APP.callback(
+            Output("lat_input", "min"),
+            Output("lat_input", "max"),
+            Output("lat_input_tooltip", "children"),
+            Output("lng_input", "min"),
+            Output("lng_input", "max"),
+            Output("lng_input_tooltip", "children"),
+            Output("map", "center"),
+            Input("location", "pathname"),
+    )
+    def initialize(path):
+        scenario = "historical"
+        model = "GFDL-ESM4"
+        variable = "pr"
+        data = xr.open_zarr(
+            f'/Data/data24/ISIMIP3b/InputData/climate/atmosphere/bias-adjusted'
```

Comment:
Extract this path to the config file. Also occurs in other places below.

Reply:
I meant to leave that out to Sheen.